### ISSUE TITLE
Overpass protocol

### ISSUE THREAD

**Issue by zstadler** [IsraelHikingMap/Site#1791]
## Feature

The [IHM custom layer](https://github.com/HarelM/IHMCustomLayerTemplate) allows everyone to define a layer to be added to IHM. However, the process is technically challenging:

1. There has to be an online geojson source for the data.
2. A MapLibre style has to be defined

Adding a protocol that receives the data from Overpass-Turbo would greatly simplify the creation of data for a custom layer. The [Overpass-Turbo wizard](https://wiki.openstreetmap.org/wiki/Overpass_turbo/Wizard) can further simplify the process. As a result, the layer data will also be kept up-to-date w.r.t. the underlying OSM data.

To address the complexity of defining a style, the protocol would have a default style. Potentially, this style will be similar to the Overpass-Turbo style. 

It would be even better if the protocol would accept simple overrides to the style, such as a URL to an icon form point features or a color for line features.

### Example Overpass-Turbo query:
- [drinking water](https://overpass-turbo.eu/s/1nyV) - following [this FB post](https://www.facebook.com/groups/israelhikingmap/posts/5598663800188767)

I could not think of a layer of line features that would be of interest to hikers or bikers. If that's so, it would be fine for the protocol to show icons only - for point and area features.

## Things I think the developers should know, images, links etc.
The code that translates Overpass API results into geojson is maintained by the Overpass-Turbo team, and it is [available on Github](https://tyrasd.github.io/osmtogeojson/) and as an npm package [osmtogeojson](https://npmjs.org/package/osmtogeojson)

For example, the above drinking water query can be defined as a layer using this syntax:
```
overpass-turbo:/1nyV
```

A possible extension could be adding formatting parameters, such as size and icon, as optional URL parameters.

**Comment by HarelM** [comment]
Reverse engineering overpass-turbo website:
1. The short link is redirected by the server to a long link: https://overpass-turbo.eu/s/1nyV --> https://overpass-turbo.eu/?Q=nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B&C=31.40054;35.04639;7&R
2. The response from the long link is an html file with the relevant query inside it as far as I can understand and javascript to run the relevant code
3. This send a post request to https://overpass-api.de/api/interpreter with the body that is inside the html: 

```
nw
  [amenity="drinking_water"]
  (area:3606195356);
out meta geom;

// rel (6195356);out geom;
```

So basically, if I'm understanding this correctly I need to do the following steps:
1. Send a call to the short URL to get the redirect address - the problem here is that I couldn't find any way to do it in javascript as this is done by the browser.
2. Take the redirect address Q parameter value (the query) and url decode it 
3. Send it using post to the interpreter
4. Translate the results into geojson using [osmtogeojson](https://npmjs.org/package/osmtogeojson)

Bottom line, I don't think this can be implemented with the short link due to how overpass works, please let me know what else is possible from your perspective as a user.

**Comment by zstadler** [comment]
Thanks for the analysis!

1. https://stackoverflow.com/a/33380848 and [responseURL doc](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseURL) suggest it is possible to capture the redirect URL in JavaScript:
   ```
   const xhr = new XMLHttpRequest();
   xhr.open('GET', 'https://overpass-turbo.eu/s/1nyV');
   xhr.onload = () => {
     console.log(xhr.responseURL);
   };
   xhr.send(null); 
   ```
   prints
   ```
   https://overpass-turbo.eu/?Q=nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B&C=31.40054;35.04639;7&R
   ```

2. The `Q` parameter can then be sent directly to the `https://overpass-api.de/api/interpreter` API as the `data` in an HTTP POST request. For example:

   ```
   curl -d 'data=nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B' https://overpass-api.de/api/interpreter
   ```

**Comment by HarelM** [comment]
I tired that, but since the overpass site doesn't support CORS the xhr request doesn't get to the `onload` but rather gets to the `onerror`, in this method the `responseURL` field is empty and I couldn't find a field in the xhr object that holds the redirected address.
Any thoughts or links to SO to solve this would help, I couldn't find anything :-(

**Comment by HarelM** [comment]
Moreover, I keep reading the following line in all the answers I read:
> To answer your question:
You can't handle redirects with XHR callbacks because the browser takes care of them automatically. You will only get back what at the redirected location.

https://stackoverflow.com/questions/15996779/cannot-handle-302-redirect-in-ajax-and-why

**Comment by HarelM** [comment]
Discussion summary: let's jump over the current problem with the short URL and focus on getting this tested assuming the user pastes the full query URL so we can see how this looks and feels...

**Comment by HarelM** [comment]
After skipping the URL redirect issue, I hard coded a lot of things and got the following to work:
I've added a source with the protocol of `overpass://some-url-i-made-up-to-see-that-I-get-it` and I've added a very simple circle layer just to make sure the data is presented:
![image](https://user-images.githubusercontent.com/3269297/203427859-55c7d3fc-5aca-4781-8717-5bb05691e16a.png)
I think a fuller less hard coded example is needed in order to better understand the manual work that is needed in order to create a custom layer.
@zstadler can you please create a style somewhere (any public github repo will do) that has a style that should be constructed as follows:
```js
...
sources: {
   ...
   overpass-example: { // or any other key you want to give it
       type: 'geojson',
       data: "overpass://query-to-overpass-url-encoded" // the example we used in this issue is nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B
   }
},
layers: [
    ... // any way you would like to style this data that uses the overpass-example source
]
```
If this will work the only question left is if we insist on the short URL or not.

**Comment by zstadler** [comment]
Here is a style file using circles similar to Overpass-Turbo: [Overpass-overlay-example.json](https://gist.github.com/zstadler/c9ce8dc115c955fde99eb6b135e0d8fe/raw/22c45cc52e0122188a11d77a78bba4b65c884eac/Overpass-overlay-example.json)

**Comment by zstadler** [comment]
[Here](https://gist.github.com/zstadler/c9ce8dc115c955fde99eb6b135e0d8fe/raw/30b4396054dd82deb2f1020502db15487819e9a1/Overpass-overlay-example.json) it is

**Comment by HarelM** [comment]
Thanks!! this is how it looks :-)
<img width="716" alt="image" src="https://user-images.githubusercontent.com/3269297/203551711-aea7af82-61e5-4b13-b5f4-8f9b3a40fb19.png">
Two things to consider
1. For some reason the layer is added twice (and removed once in between) which I need to investigate in order to improve performance and avoid sending too many requests to overpass
2. What to do about the short URL.

This is less than 15 lines of code in total :-)

**Comment by zstadler** [comment]
There is a free [un-shorten API - https://unshorten.me/api](https://unshorten.me/api).

For example, [`https://unshorten.me/s/https://overpass-turbo.eu/s/1nyV`](https://unshorten.me/s/https://overpass-turbo.eu/s/1nyV) returns
```
https://overpass-turbo.eu/?Q=nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B&C=31.40054;35.04639;7&R
```

**Comment by zstadler** [comment]
One other thing to consider is the Overpass-Turbo feature that prevents small features from vanishing when zooming-out. 

Overpass Turbo seems to do so by temporarily replacing the geometry of the feature with a point geometry when the size of the bounding box is below a threshold. The original geometry is restored when the appropriate zoom-in is done.

[Their code](https://github.com/tyrasd/overpass-turbo/blob/6d05bf146378d752839397ee3e7dbc290c20c0d4/js/GeoJsonNoVanish.js#L23-L74) is made for Leaflet.

**Comment by HarelM** [comment]
The above doesn't support CORS unfortunately... :/
```
Access to XMLHttpRequest at 'https://unshorten.me/s/https://overpass-turbo.eu/s/1nyV' from origin 'http://localhost:5000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

In the style users are free to define min zoom, max zoom etc and style however they see fit, I don't think there's a need here to complicate things...

**Comment by HarelM** [comment]
Bottom line, at this point I have no more intentions on pursuing the short URL issue.
The style editor should take the query, url encode it using an online tool like https://www.urlencoder.org/, add a prefix of `overpass://` and set that as the value for the `data` field in the geojson source, much like can be see in the following example from [here](https://gist.githubusercontent.com/zstadler/c9ce8dc115c955fde99eb6b135e0d8fe/raw/30b4396054dd82deb2f1020502db15487819e9a1/Overpass-overlay-example.json):
```json
{
  "version": 8,
  "name": "Overpass-overlay-example",
  "metadata": {
    "owner": "Zeev Stadler",
    "copyright": "Copyright Zeev Stadler 2022. All Rights Reserved."
  },
  "sources": {
    "overpass-points": {
      "type": "geojson",
      "data": "overpass://nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B"
    }
  },
  "layers": [
    {
      "id": "overpass-point",
      "type": "circle",
      "metadata": {"IHM:overlay": true},
      "source": "overpass-points",
      "paint": {
        "circle-radius": 10,
        "circle-color": "#f22",
        "circle-stroke-width": 2,
        "circle-opacity": 0.3,
        "circle-stroke-color": "#d0f",
        "circle-stroke-opacity": 0.7
      }
    }
  ]
}
```

**Comment by zstadler** [comment]
I've added a proxy for the unshorten API at https://israelhiking.osm.org.il/unshorten

For example, the body of the response to https://israelhiking.osm.org.il/unshorten/overpass-turbo.eu/s/1nyV is
```
https://overpass-turbo.eu/?Q=nw%0A%20%20%5Bamenity%3D%22drinking_water%22%5D%0A%20%20(area%3A3606195356)%3B%0Aout%20meta%20geom%3B%0A%0A%2F%2F%20rel%20(6195356)%3Bout%20geom%3B&C=31.40054;35.04639;7&R
```

![image](https://user-images.githubusercontent.com/1304610/203965706-92aeb979-df8c-4ff0-9af5-f7ccc68f3354.png)

**Comment by zstadler** [comment]
Some Overpass-Turbo query links are LZW-compressed, as noted by a `?q=` parameter rather than the `?Q=` parameter used for plain text queries.

For example, https://overpass-turbo.eu/s/1ocA is a link to same "drinking water" query above, but it is compressed.
The body of the response to https://israelhiking.osm.org.il/unshorten/overpass-turbo.eu/s/1ocA is
```
https://overpass-turbo.eu/?q=bncKICBbYW1lbml0eT0iZHJpbmvEkWdfd2F0ZXIiXcSCIChhcmVhOjM2MDYxOTUzNTYpOwpvdXQgxId0YSBnZW9txLEKLy8gxKJsxJ_EqcSrxK3ErzvEs8S1xLvEvTs&c=BYAznW9WWH&R
```

Luckily, the de-compression code is written in JavaScript and it is available in the Overpass-Turbo repository:
- [parameter handling](https://github.com/tyrasd/overpass-turbo/blob/febbccd5a2cc4f089fac3f55963d3262e45f81ec/js/urlParameters.js#L22-L31)
- [query decompression](https://github.com/tyrasd/overpass-turbo/blob/febbccd5a2cc4f089fac3f55963d3262e45f81ec/js/misc.js#L240-L263)

**Comment by HarelM** [comment]
Using the style.json. overpass is just the underlying data source.

**Comment by amiad** [comment]
I tried to add a layer but I must have made a mistake. I'm not a geojson expert.

https://gist.github.com/amiad/e4e7187e906a716ec279fdd0bb28b051

**Comment by zstadler** [comment]
@amiad,

[Maputnik](https://maputnik.github.io/editor) is your friend.

![image](https://github.com/IsraelHikingMap/Site/assets/1304610/2f05c321-cc94-4cc0-bf58-208cb42d4d9d)

- "Open" allows you to load a style from a file or a URL
  ![image](https://github.com/IsraelHikingMap/Site/assets/1304610/21907890-0495-4b94-ac54-3aefb8e52f52)
- "Data Sources" allows you to use a copy of the GeoJSON from Overpass-Turbo as a temporary data source
  ![image](https://github.com/IsraelHikingMap/Site/assets/1304610/876e3470-d48e-4ce7-82ae-61a3ebe23660)
- Provides useful error messages
  ![image](https://github.com/IsraelHikingMap/Site/assets/1304610/77a2cc23-ff42-4fff-b470-fa704d6127dd)
- Provides a list of available fonts
  ![image](https://github.com/IsraelHikingMap/Site/assets/1304610/68de098c-e96b-42f9-a347-8d45eef41570)
- More...

Also, `out geom;` would be sufficient.

**Comment by amiad** [comment]
תודה, אחרי העזרה ממך ומהראל הצלחתי להגדיר הכול כמו שרציתי.

![תמונה](https://github.com/IsraelHikingMap/Site/assets/402108/f27f7653-089f-429d-b6ea-dfe42317891a)

**Comment by zstadler** [comment]
Great!

Would you be interested in publishing your work

1. As an [additional overlay](https://github.com/IsraelHikingMap/Site/wiki/Custom-maps#%D7%A9%D7%9B%D7%91%D7%95%D7%AA-%D7%A9%D7%A7%D7%95%D7%A4%D7%95%D7%AA), similar to the waterhole status overlay?
2. As a new [Wiki page](https://github.com/IsraelHikingMap/Site/wiki) for overpass-based overlays?

**Comment by amiad** [comment]
כן, רק אני לא בטוח מה העניין לציבור. המטרה האישית שלי היא להציג את סוג העץ גם במקרה שהוא לא קיים ב־name (או שמשתמש חרוץ מחק אותו משם)
